dashboard/app/models/user.rb

```diff
@@ -1138,6 +1138,46 @@ def has_activity?
     user_levels.attempted.exists?
   end
 
+  # Returns the next visible script_level for the next progression level in the given
+  # script that hasn't yet been passed, starting its search at the last level we submitted
+  def next_unpassed_visible_progression_level(script)
+    # visible script levels
+    sl_level_ids = visible_script_levels(script).map(&:level_ids).flatten
+
+    # levels the user made progress in
+    ul_level_ids = user_levels_by_level(script).keys
+
+    # The user has not made any progress, return the first script_level
+    if ul_level_ids.empty?
+      return script.get_script_level_by_chapter(1)
+    end
+
+    visible_completed_level_ids = sl_level_ids & ul_level_ids
+
+    # Find the user_levels associated with visible script_levels
+    visible_user_levels = user_levels.where(level_id: visible_completed_level_ids)
+
+    # Most recently completed user_level of the visible subset
+    most_recent_ul = visible_user_levels.max_by(&:created_at)
+
+    # Find the script_level that goes with the most recent user_level
+    most_recent_sl = visible_script_levels(script).detect do |sl|
+      sl.level_id == most_recent_ul.level_id
+    end
+
+    # Find the chapter for the script_level that goes with the most recent user_level
+    most_recent_completed_chapter = most_recent_sl.chapter
+
+    # Find the script_level that has the next highest chapter level from the one above
+    later_visible_sls = visible_script_levels(script).select do |sl|
+      sl.chapter > most_recent_completed_chapter
+    end
+
+    next_unpassed_visible_progression_sl = later_visible_sls.min_by(&:chapter)
+
+    next_unpassed_visible_progression_sl
```

Comment:
does there need to be fallback logic here, akin to https://github.com/code-dot-org/code-dot-org/blob/9be5d1e0c0752205526ea796045c7b7d358919e2/dashboard/app/models/user.rb#L1162-L1164 ?

Reply:
With the latest changes I handle this logic within the `next_unpassed_visible_progression_level` method, which is even more of a beast so I don't know if that was the best decision...